message_edit: Refactor message edit code.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First two commits are refactor commits.
- Last commit is the main refactor commit.

 <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on zerver/actions/message_edit.py:
```diff
@@ -526,67 +633,21 @@ def do_update_message(
-            info.stream_push_user_ids,
-        )
+            changed_messages_count = 1
+            return UpdateMessageResult(
```
Did this change, so that we can just return early for direct messages and we can easily handle different `message_edit_request` types for stream and direct messages.

sahil839 commented on zerver/actions/message_edit.py:
```diff
@@ -1296,6 +1268,83 @@ def check_time_limit_for_change_all_propagate_mode(
+    topic_resolved = False
+    topic_unresolved = False
+    old_topic_name = message.topic_name()
+    new_topic_name = old_topic_name
```
Have kept this so that `topic_name` and `stream` can be set to orginal value if they are not changed and the new value otherwise so that we do not need conditions like `stream = new_stream if new_stream is not None else orig_stream`.
And we have boolean fields for checking which of the things were changed.

sahil839 commented on zerver/actions/message_edit.py:
```diff
@@ -1296,6 +1268,83 @@ def check_time_limit_for_change_all_propagate_mode(
+    is_stream_edited = False
+    new_stream = orig_stream
+    if stream_id is not None:
+        new_stream = access_stream_by_id(user_profile, stream_id, require_active=True)[0]
```
Fetching the stream here means that validation of the new stream ID is done here and invalid or inaccessible stream ID error would be raised first before other errors, which was not the case before. But this seems better than just setting the ID here and then have the `stream` field optional and then using assert statements to make sure that the `stream` field is not None.

sahil839 commented on zerver/lib/types.py:
```diff
@@ -354,3 +357,28 @@ class AnalyticsDataUploadLevel(IntEnum):
+
+
+@dataclass
+class DirectMessageEditRequest:
```
Added two different types as direct messages do not have stream and topic related fields and we would need to keep those as optional for them which will require assert statements in the code to keep mypy happy.

Also, it was suggested to have different classes for content edit, topic edit and stream edit and then check the types, but since we allow changing multiple things in a single request, I went ahead with a single `StreamMessageEditRequest` class.

timabbott replied:
I don't think we still allow changing content at the same time as anything else. So I think there are probably 6 types:
- Content edit for DM
- Content edit for channel message
- Resolve topic
- Change just topic for channel message
- Change just channel for channel message
- Change both topic and channel for channel message.

sahil839 replied:
API still allows allow topic and content together.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -430,6 +430,111 @@ def get_visibility_policy_after_merge(
+
+    With topic edits, propagate_mode determines whether other message
+    also have their topics edited.
+    """
```
This docstring seems misplaced, if this the content code path?

sahil839 replied:
Fixed this.

timabbott commented on zerver/tests/test_message_fetch.py:
```diff
@@ -4988,15 +4988,15 @@ def update_message(self, msg: Message, content: str) -> None:
         mention_backend = MentionBackend(realm_id)
         mention_data = MentionData(mention_backend, content, msg.sender)
+        message_edit_request = get_message_edit_request_object(
+            msg, hamlet, "change_one", None, None, content
```
I'd prefer to pass all the parameter as mandatory kwargs, not having any positional parameters, for `get_message_edit_request_object`; it'd be a lot more readable.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -598,20 +576,15 @@ def user_info(um: UserMessage) -> dict[str, Any]:
-            stream_id=stream_id,
-            topic_name=new_topic_name,
+            stream_id=message_edit_request.orig_stream.id,
+            topic_name=message_edit_request.topic_name,
```
Why is this a combination of the original stream and new topic?

sahil839 replied:
This is only used for content edits and we do not allow editing stream with content. I too was confused about this while changing the code.

Maybe we can move this in `if is_content_edited` block.

timabbott replied:
Yeah, clarifying that, maybe in a follow-up commit, seems good.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -1298,6 +1270,85 @@ def check_time_limit_for_change_all_propagate_mode(
+        orig_topic_name=old_topic_name,
+        orig_stream=orig_stream,
+        propagate_mode=propagate_mode,
+        stream=new_stream,
```
I think `new_stream` or `target_stream` would be a better name in the dataclass; when we have two streams to talk about here, it's most readable if neither has bare `stream` as the name. Same with `topic_name`.

sahil839 replied:
`stream` field is set to the original stream if stream is not changed. This is essentially to avoid passing arguments like like `stream=message_edit_request.new_stream if message_edit_request.new_stream is not None else message_edit_request.orig_stream`. Similarly for topics as well.

So, `target_stream` and `target_topic_name` might be better name?

timabbott replied:
I think `target_` seems clear, yeah.

timabbott commented:
This looks broadly good; posted a couple comments on naming.

sahil839 commented:
Updated the PR -
- We use `target_stream` and `target_topic_name` fields now.
- Also, added a commit for the `StreamTopic` value passed when editing content.

`StreamMessageEditRequest` still uses `content` field, since `target_content` sounds weird. Can name it as `new_content` though and keep it `None` if content is not changed.

timabbott commented:
OK, this seems pretty good, merged, thanks @sahil839!

timabbott commented:
`get_message_edit_request_object` might be better named `build_message_edit_request`.
